### ISSUE TITLE
[ONNX] small eps of LayerNorm will be exported as 0.0 in fp16 mode

### ISSUE THREAD

**Issue by mindest** [pytorch/pytorch#86580]
### 🐛 Describe the bug

For `torch.nn.LayerNorm` in fp16 mode, when `eps` is smaller than `2^(-24)` (minimal fp16 positive number), it will be exported as a constant `0.0`. This is different from PyTorch behavior - it still keeps `eps` and necessary calculations as float in fp16 mode. Is it possible to fix this like adding casts in the beginning (to float) and at the end (back to fp16) of the [export function](https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset9.py#L2661-L2689) when running in fp16 mode, since some computations needed to be done in fp32? Thanks.

Minimal export case:
```py
import torch

class Net(torch.nn.Module):
    def __init__(self, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm(C, eps=1e-8)
    def forward(self, x):
        return self.layer_norm(x)

N, C = 8, 4
model = Net(C).cuda().half()
x = torch.randn(N, C).cuda().half()

torch.onnx.export(model, x, "test_layernorm_export_fp16.onnx", opset_version=12)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.0.0-1032-azure-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] pytorch-lightning==1.6.3
[pip3] torch==1.11.0
[pip3] torch-nebula==0.15.1
[pip3] torch-ort==1.11.0
[pip3] torchmetrics==0.7.1
[pip3] torchvision==0.12.0+cu113
[conda] magma-cuda113             2.5.2                         1    pytorch
[conda] mkl                       2021.4.0                 pypi_0    pypi
[conda] mkl-include               2021.4.0                 pypi_0    pypi
[conda] numpy                     1.22.2                   pypi_0    pypi
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torch-nebula              0.15.1                   pypi_0    pypi
[conda] torch-ort                 1.11.0                   pypi_0    pypi
[conda] torchmetrics              0.7.1                    pypi_0    pypi
[conda] torchvision               0.12.0+cu113             pypi_0    pypi
```

**Comment by AllenTiTaiWang** [comment]
If this is blocking you, could you try with nightly PyTorch and opset_version 17?  In 17, ONNX specifically supports [LayerNorm](https://github.com/onnx/onnx/blob/main/docs/Operators.md#LayerNormalization). Before 17, we convert aten::layer_norm with other ONNX nodes, which we will need more time to address this.

**Comment by mindest** [comment]
> If this is blocking you, could you try with nightly PyTorch and opset_version 17? In 17, ONNX specifically supports [LayerNorm](https://github.com/onnx/onnx/blob/main/docs/Operators.md#LayerNormalization). LMK if the node matches the torch behavior. Before 17, like you point out, we convert aten::layer_norm with other ONNX nodes. though it seems doable, we still need more time to look into this.

Thanks @AllenTiTaiWang. Currently we are using opset 15 and can workaround by modifying model code to use larger `eps`. Still it would be good if we can avoid modifying user code when using opest < 17.

**Comment by AllenTiTaiWang** [comment]
Before 17:
```bash
ONNX :  graph(%input : Half(8, 4, strides=[4, 1], requires_grad=0, device=cuda:0),
      %layer_norm.weight : Half(4, strides=[1], requires_grad=1, device=cuda:0),
      %layer_norm.bias : Half(4, strides=[1], requires_grad=1, device=cuda:0)):
  %6 : Double(device=cpu) = onnx::Constant[value={2}](), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %7 : Double(device=cpu) = onnx::Constant[value={1e-08}](), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %8 : Half(8, 1, strides=[1, 1], device=cpu) = onnx::ReduceMean[axes=[-1]](%input), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %9 : Half(8, 4, strides=[4, 1], device=cpu) = onnx::Sub(%input, %8), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %10 : Half(8, 4, strides=[4, 1], device=cpu) = onnx::Pow(%9, %6), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %11 : Half(8, 1, strides=[1, 1], device=cpu) = onnx::ReduceMean[axes=[-1]](%10), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %12 : Half(8, 1, strides=[1, 1], device=cpu) = onnx::Add(%11, %7), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %13 : Half(8, 1, strides=[1, 1], device=cpu) = onnx::Sqrt(%12), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %14 : Half(8, 4, strides=[4, 1], device=cpu) = onnx::Div(%9, %13), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %15 : Half(8, 4, strides=[4, 1], device=cpu) = onnx::Mul(%14, %layer_norm.weight), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  %16 : Half(8, 4, strides=[4, 1], requires_grad=1, device=cuda:0) = onnx::Add(%15, %layer_norm.bias), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  return (%16)
```

Although eps is merely a parameter to layer_norm, the lack of this ONNX op before 17 needs us process it as input/output during exporting, which causes the issue. However, changing this would be touching shape type inference, which needs more time to work on.

If there are others blocked by this, opset 17 should be working as it's converting like:
```bash
ONNX :  graph(%input : Half(8, 4, strides=[4, 1], requires_grad=0, device=cuda:0),
      %layer_norm.weight : Half(4, strides=[1], requires_grad=1, device=cuda:0),
      %layer_norm.bias : Half(4, strides=[1], requires_grad=1, device=cuda:0)):
  %6 : Half(8, 4, strides=[4, 1], requires_grad=1, device=cuda:0) = onnx::LayerNormalization[axis=-1, epsilon=1e-08](%input, %layer_norm.weight, %layer_norm.bias), scope: __main__.Net::/torch.nn.modules.normalization.LayerNorm::layer_norm # /home/titaiwang/pytorch/torch/nn/functional.py:2515:0
  return (%6)
```

**Comment by AllenTiTaiWang** [comment]
@mindest 
Could you provide me one example that this behavior making prediction mismatch?
I thought about fixing type during eps calculations, but can't find a case causing mismatch.

**Comment by mindest** [comment]
@AllenTiTaiWang The LayerNorm op is exported as several operators, which could have precision loss when calculating in fp16 mode (like ReduceMean) and should be left in fp32. I think this could be the reason for our model's AUC loss. But I didn't look into details exactly where the loss is from.

**Comment by titaiwangms** [comment]
@mindest Could you try the same easy fix on my PR with your local PyTorch build to verify it works? I think it makes sense keeping Double before `Div` and after `Add` on eps, but I am not sure if this is satisfactory.

**Comment by mindest** [comment]
@titaiwangms I think this change will introduce more extra casts (to float64) than necessary in the exported graph, and it will change the output dtype to float64, too. Would adding two casts before and after layer_norm calculation (like the following) be enough to handle half precision case? Here I don't consider extra outputs like `rdenominator` etc.

```diff
    # ...
+   is_input_half = input.type().scalarType() == 'Half'
+   if is_input_half:
+       input = g.op("Cast", input, to_i=symbolic_helper.cast_pytorch_to_onnx["Float"])
    mean = g.op("ReduceMean", input, axes_i=axes)
    numerator = sub(g, input, mean)
    # variance = e((x - e(x))^2), and (x - e(x)) is the numerator in the layer_norm formula
    variance = g.op("ReduceMean", pow(g, numerator, two_cst), axes_i=axes)
    denominator = sqrt(g, add(g, variance, eps_cst))

    normalized = g.op("Div", numerator, denominator)
+   if is_input_half:
+       normalized = g.op("Cast", normalized, to_i=symbolic_helper.cast_pytorch_to_onnx["Half"])
    # ...
```
